PLFM-2073 Phase 1 - Invite a new user to join a team

hhu94 opened the pull request:
This PR implements the remaining part of phase 1 of the project, which includes the following tasks:

Create new models
- [x] InviteeVerificationSignedToken

Update existing models
- [x] MembershipInvtnSubmission

Update existing services
- [x] POST /membershipInvitation

Create new services
- [x] POST /membershipInvitation/{id}
- [x] POST /membershipInvitation/{id}/verification
- [x] PUT /membershipInvitation/{id}/inviteeId

Commit:
Implement changes to POST /membershipInvitation

Commit:
Fix typo

Commit:
Temporarily make etag nullable

Commit:
Fix service test

Commit:
Move team related models to their own directory

Commit:
Revert "Move team related models to their own directory"

This reverts commit ae77b2871329fedc085036547a8f1497674423c2.

Commit:
Merge branch 'develop' into PLFM-2073

Commit:
Implement /membershipInvitation/{id}/verification

Commit:
Add verifyInvitee to Java client

Also, add integration test for it.
verifyInvitee sends a POST /membershipInvitation/{id}/verification.

Commit:
Merge branch 'develop' into PLFM-2073

Commit:
Add service to get invitation with signed token

POST /membershipInvitation/{id} with MembershipInvtnSignedToken as
request body

Commit:
Implement PUT /membershipInvitation/{id}/inviteeId

Commit:
Add verifyInvitee to Java client

Commit:
Add overloaded getInvitation to Java client

This overloaded method uses a MembershipInvtnSignedToken for
authorization

Commit:
Add comments and other small changes

Edit comments, delete unused imports and rename a method

Commit:
Move updateInviteeId logic from DAO to manager

Also update tests to reflect the change

Commit:
Update etag when inviteeId is updated

brucehoff commented on services/repository/src/main/java/org/sagebionetworks/repo/web/controller/MembershipInvitationController.java:
```diff
@@ -165,4 +177,41 @@ public void deleteInvitation(
+	 * Verify whether the inviteeEmail of the indicated MembershipInvitation is associated with the authenticated user.
+	 * If it is, the response body will contain an InviteeVerificationSignedToken.
+	 * If it is not, the response body will be null.
+	 * @param membershipInvitationId
```
Please consider adding a link to the design document.

brucehoff replied:
An alternative pattern:  let this service be an HTTP GET which returns the InviteeVerificationSignedToken (status=200 OK) if the authenticated user is associated with the embedded email or 403 Unathorized if the user is anonymous or if the embedded email address is NOT associated with the authenticated user.

brucehoff commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/MembershipInvitationManagerImpl.java:
```diff
@@ -94,7 +93,7 @@ public MessageToUserAndBody createInvitationNotification(MembershipInvtnSubmissi
 		fieldValues.put(TEMPLATE_KEY_TEAM_NAME, teamDAO.get(mis.getTeamId()).getName());
 		fieldValues.put(TEMPLATE_KEY_TEAM_ID, mis.getTeamId());
-		fieldValues.put(TEMPLATE_KEY_ONE_CLICK_JOIN, EmailUtils.createOneClickJoinTeamLink(
+		fieldValues.put(TEMPLATE_KEY_ONE_CLICK_JOIN, EmailUtils.createMembershipInvtnLink(
```
this change may be wrong

hhu94 replied:
Fixed

brucehoff commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/MembershipInvitationManagerImpl.java:
```diff
@@ -162,6 +192,49 @@ public Count getOpenInvitationCountForUser(String principalId) {
+		ValidateArgument.required(userId, "userId");
+		ValidateArgument.required(membershipInvitationId, "membershipInvitationId");
+		// Get list of email addresses associated with userId
+		List<PrincipalAlias> aliases = principalAliasDAO.listPrincipalAliases(userId);
```
instead of retrieving all aliases, call a DAO-level method that just checks whether the given user has the alias of interest

brucehoff commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/MembershipInvitationManagerImpl.java:
```diff
@@ -162,6 +192,49 @@ public Count getOpenInvitationCountForUser(String principalId) {
+		}
+		MembershipInvtnSubmission mis = membershipInvtnSubmissionDAO.get(misId);
+		if (!(mis.getInviteeId() == null && mis.getInviteeEmail() != null)) {
+			throw new UnauthorizedException("Cannot update inviteeId. The target invitation must have a null inviteeId and a non null inviteeEmail.");
```
This case should return a 400 level status. (IllegalArgumentException)

brucehoff commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOMembershipInvtnSubmissionDAOImpl.java:
```diff
@@ -292,5 +292,20 @@ public void deleteByTeamAndUser(long teamId, long inviteeId)
+		param.addValue(COL_MEMBERSHIP_INVITATION_SUBMISSION_INVITEE_ID, inviteeId);
+		param.addValue(COL_MEMBERSHIP_INVITATION_SUBMISSION_ID, misId);
+		param.addValue(COL_MEMBERSHIP_INVITATION_SUBMISSION_ETAG, UUID.randomUUID().toString());
+		return namedJdbcTemplate.update(UPDATE_INVITEE_ID, param);
```
return type -> void
if number updated == 0 throw NotFoundException (misId is not found) or IllegalArgumentException.
if number updated > 1 then throw DatastoreException, which maps to a 500 status.

brucehoff commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/MembershipInvitationManagerImpl.java:
```diff
@@ -162,6 +192,49 @@ public Count getOpenInvitationCountForUser(String principalId) {
+		}
+		int rowsUpdated = membershipInvtnSubmissionDAO.updateInviteeId(misId, userId);
+		if (rowsUpdated != 1) {
+			throw new DatastoreException("Expected to update 1 row, but instead updated " + rowsUpdated + " rows.");
```
this check should be in the DAO level

brucehoff commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/MembershipRequestManagerImpl.java:
```diff
@@ -139,7 +137,7 @@ public MembershipRqstSubmission create(UserInfo userInfo,
 		Set<String> teamAdmins = new HashSet<String>(teamDAO.getAdminTeamMembers(mrs.getTeamId()));
 		for (String recipientPrincipalId : teamAdmins) {
-			fieldValues.put(TEMPLATE_KEY_ONE_CLICK_JOIN, EmailUtils.createOneClickJoinTeamLink(
+			fieldValues.put(TEMPLATE_KEY_ONE_CLICK_JOIN, EmailUtils.createMembershipInvtnLink(
```
mistake

hhu94 replied:
Fixed

brucehoff commented on services/repository-managers/src/main/resources/message/emailTeamMembershipInvitationExtendedTemplate.html:
```diff
@@ -0,0 +1,24 @@
+<html style="-webkit-box-sizing: border-box;-moz-box-sizing: border-box;box-sizing: border-box;font-size: 10px;-webkit-tap-highlight-color: rgba(0, 0, 0, 0);">
```
consider changing name to teamMembershipInvitationExtenededToEmailTemplate.html.  This makes the file list next to the other variation.

hhu94 replied:
Fixed

brucehoff commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/MembershipInvitationManagerImpl.java:
```diff
@@ -162,6 +192,49 @@ public Count getOpenInvitationCountForUser(String principalId) {
+		if (!status.getAuthorized()) {
+			throw new UnauthorizedException(status.getReason());
+		}
+		MembershipInvtnSubmission mis = membershipInvtnSubmissionDAO.get(misId);
```
this should be a select for update to ensure that no other thread changes the record before 'updateInviteeid' is callsed, below.

brucehoff commented on client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClient.java:
```diff
@@ -1593,7 +1554,27 @@ MembershipInvtnSubmission createMembershipInvitation(
+	 * Verify whether the inviteeEmail of the indicated MembershipInvitation is associated with the authenticated user.
+	 * If it is, the response body will contain an InviteeVerificationSignedToken.
+	 * If it is not, the response body will be null.
+	 * @param membershipInvitationId
```
Note:  This will update if the controller is changed.

hhu94 replied:
Fixed

brucehoff replied:
"getInviteeVerificationSignedToken()"

brucehoff commented on integration-test/src/test/java/org/sagebionetworks/IT500SynapseJavaClient.java:
```diff
@@ -1426,6 +1389,32 @@ public void testMembershipInvitationAPI() throws Exception {
+		assertEquals(mis.getId(), token.getMembershipInvitationId());
+
+		// update the inviteeIinviteeUserProfile.getOwnerId()d of the invitation
+		synapseTwo.updateInviteeId(mis.getId(), token);
```
typo

hhu94 replied:
Fixed

brucehoff commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthorizationManagerImpl.java:
```diff
@@ -608,4 +609,30 @@ public AuthorizationStatus canAccessMembershipInvitationSubmission(UserInfo user
+
+	@Override
+	public AuthorizationStatus canAccessMembershipInvitationSubmission(String misId, MembershipInvtnSignedToken token, ACCESS_TYPE accessType) {
+		try {
```
omit 'misId' since the ID is already in the token.

hhu94 replied:
Fixed

brucehoff commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthorizationManagerImpl.java:
```diff
@@ -608,4 +609,30 @@ public AuthorizationStatus canAccessMembershipInvitationSubmission(UserInfo user
+	}
+
+	@Override
+	public AuthorizationStatus canAccessMembershipInvitationSubmission(Long userId, String misId, InviteeVerificationSignedToken token, ACCESS_TYPE accessType) {
```
please omit misId, which is redundant wrt the token content

hhu94 replied:
Fixed

brucehoff commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/EmailUtils.java:
```diff
@@ -202,7 +203,7 @@ public static String createSource(String senderDisplayName, String senderUserNam
 
 	
-	public static String createOneClickJoinTeamLink(String endpoint, String userId, String memberId, String teamId, Date createdOn) {
+	public static String createMembershipInvtnLink(String endpoint, String userId, String memberId, String teamId, Date createdOn) {
```
didn't mean to change this name

hhu94 replied:
Fixed

Commit:
Undo wrong refactor of createOneClickJoinTeamLink

Commit:
Add expiresOn to MembershipInvtnSignedToken

Change signature of verifyInvitee to include MembershipInvtnSignedToken
Check expiresOn of the token passed into verifyInvitee

Commit:
Rename template and fix typo

Commit:
Move ID consistency check to MIManager

Commit:
Throw UnauthorizedException in verifyInvitee

Commit:
Fix testUpdateId

Commit:
Avoid potential race condition in updateInviteeId

Commit:
Change expiration behavior of MISignedToken

Commit:
Fix test

Commit:
Add expiresOn to InviteeVerificationSignedToken

Set a 24 hour expiration period on tokens created by verifyInvitee
Enforce expiresOn check on the token passed into updateInviteeId
Edit comments to reflect expiresOn changes

Commit:
Add description to expiresOn

brucehoff commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOMembershipInvtnSubmissionDAOImpl.java:
```diff
@@ -143,6 +142,16 @@ public MembershipInvtnSubmission get(String id) throws DatastoreException, NotFo
 		return dto;
 	}
 
+	@WriteTransactionReadCommitted
```
I don't think this has any effect.  the transaction is established at the manager level.

brucehoff commented on lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/MembershipInvtnSignedToken.json:
```diff
@@ -10,6 +10,11 @@
     "membershipInvitationId": {
       "type": "string",
       "description": "The ID of the membership invitation."
+    },
```
as we discussed, this is not needed, being redundant w.r.t. the expiration in the invitation

brucehoff commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/MembershipInvitationManagerImpl.java:
```diff
@@ -193,46 +196,50 @@ public Count getOpenInvitationCountForUser(String principalId) {
-			SignedTokenUtil.signToken(token);
-			return token;
+		// Find out if the given user has an alias that matches the inviteeEmail
+		PrincipalAlias principalAlias = principalAliasDAO.findPrincipalWithAlias(inviteeEmail);
```
I would like to see a test for the case in which the email associated with the user is "xxx.com" and the email entered in the invitation is "xxxx.COM", i.e. the normalized alias matches but the alias entered by the users does not.

brucehoff commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/MembershipInvitationManagerImpl.java:
```diff
@@ -193,46 +196,50 @@ public Count getOpenInvitationCountForUser(String principalId) {
+		InviteeVerificationSignedToken response = new InviteeVerificationSignedToken();
+		response.setInviteeId(userId.toString());
+		response.setMembershipInvitationId(membershipInvitationId);
+		response.setExpiresOn(new Date(new Date().getTime() + TWENTY_FOUR_HOURS_IN_MS));
```
new Date().getTime() == System.currentTimeMillis()

Commit:
Remove expiresOn from MISignedToken

Add aliasIsBoundToPrincipal to PrincipalAliasDAOImpl, and
use it in verifyInvitee instead of findPrincipalWithAlias

brucehoff commented on client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClientImpl.java:
```diff
@@ -4043,7 +4044,7 @@ public Count getOpenMembershipInvitationCount() throws SynapseException {
 	@Override
 	public InviteeVerificationSignedToken verifyInvitee(String membershipInvitationId) throws SynapseException {
 		String uri = MEMBERSHIP_INVITATION + "/" + membershipInvitationId + "/verification";
```
can the URI be something like /membershipInvitation/{membershipInvitationId}/inviteeVerificationSignedToken

brucehoff commented on integration-test/src/test/java/org/sagebionetworks/IT500SynapseJavaClient.java:
```diff
@@ -1417,6 +1417,14 @@ public void testMembershipInvitationAPI() throws Exception {
 		}
 	}
 
+	private String parseMembershipInvtnSignedToken(String invitationEmail) {
```
unused code

brucehoff commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/MembershipInvitationManagerImpl.java:
```diff
@@ -196,43 +195,43 @@ public Count getOpenInvitationCountForUser(String principalId) {
+		if (!misId.equals(token.getMembershipInvitationId())) {
+			throw new IllegalArgumentException("ID in URI and ID in signed token don't match");
+		}
+		if (token.getExpiresOn() != null && token.getExpiresOn().before(new Date())) {
```
if (token.getExpiresOn())==null) throw new IAE("expiration is missing.");
if (token.getExpiresOn().before(new Date())) {) throw new IllegalArgumentException("InviteeVerificationSignedToken is expired");
